examples/testing-websockets/src/main.rs

```diff
@@ -0,0 +1,149 @@
+            if socket
+                .send(Message::Text(format!("You said: {msg}")))
+                .await
+                .is_err()
+            {
+                break;
+            }
+        }
+    }
+}
+
+// The unit testable version requires some changes.
+//
+// By splitting the socket into an `impl Sink` and `impl Stream` we can test without providing a
+// real socket and instead using channels, which also implement `Sink` and `Stream`.
+async fn unit_testable_handler(ws: WebSocketUpgrade) -> Response {
+    ws.on_upgrade(|socket| {
+        let (write, read) = socket.split();
+        unit_testable_handle_socket(write, read)
+    })
+}
+
+// The implementation is largely the same as `integration_testable_handle_socket` expect we call
+// methods from `SinkExt` and `StreamExt`.
+async fn unit_testable_handle_socket<W, R>(mut write: W, mut read: R)
+where
+    W: Sink<Message> + Unpin,
+    R: Stream<Item = Result<Message, axum::Error>> + Unpin,
+{
+    while let Some(Ok(msg)) = read.next().await {
+        if let Message::Text(msg) = msg {
+            if write
+                .send(Message::Text(format!("You said: {msg}")))
+                .await
+                .is_err()
+            {
+                break;
+            }
+        }
+    }
+}
+
+#[cfg(test)]
+mod tests {
+    use super::*;
+    use std::net::Ipv4Addr;
+    use tokio_tungstenite::tungstenite;
+
+    // We can integration test one handler by running the server in a background task and
+    // connecting to it like any other client would.
+    #[tokio::test]
+    async fn integration_test() {
+        let server = axum::Server::bind(&SocketAddr::from((Ipv4Addr::UNSPECIFIED, 0)))
+            .serve(app().into_make_service());
+        let addr = server.local_addr();
+        tokio::spawn(server);
+
+        let (mut socket, _response) =
+            tokio_tungstenite::connect_async(format!("ws://{addr}/integration-testable"))
+                .await
+                .unwrap();
+
+        socket
+            .send(tungstenite::Message::text("foo"))
+            .await
+            .unwrap();
+
+        let msg = match socket.next().await.unwrap().unwrap() {
+            tungstenite::Message::Text(msg) => msg,
+            other => panic!("expected a text message but got {other:?}"),
+        };
+
+        assert_eq!(msg, "You said: foo");
+    }
+
+    // We can unit test the other handler by creating channels to read and write from.
+    #[tokio::test]
+    async fn unit_test() {
+        // Need to use "futures" channels rather than "tokio" channels as they implement `Sink` and
+        // `Stream`
+        let (socket_write, mut test_rx) = futures::channel::mpsc::channel(1024);
+        let (mut test_tx, socket_read) = futures::channel::mpsc::channel(1024);
+
+        tokio::spawn(unit_testable_handle_socket(socket_write, socket_read));
+
+        test_tx
+            .send(Ok(Message::Text("foo".to_owned())))
+            .await
+            .unwrap();
```

Comment:
Can be improved slightly like:

```rust
let (socket_write, mut test_rx) = futures::channel::mpsc::channel(1024);

let msgs = vec![Message::Text("foo".to_owned())];
let msg_stream = stream::iter(msgs.into_iter().map(Ok));

tokio::spawn(unit_testable_handle_socket(socket_write, msg_stream));
```

Reply:
I kinda wanna keep it simple as I think thats easier for people to adapt to their needs 🤔

Reply:
Both patterns look useful. This one is more convenient for multiple messages sent at once.

Reply:
This pattern can be added separately or we can keep it simple and omit it. Either looks fine.